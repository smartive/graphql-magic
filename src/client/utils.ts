import upperCase from 'lodash/upperCase';

export const constantCase = (str: string) => upperCase(str).replace(/ /g, '_');
