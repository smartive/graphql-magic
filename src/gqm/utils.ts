import { SourceFile, SyntaxKind, SyntaxList } from 'ts-morph';

export const findDeclarationInFile = (sourceFile: SourceFile, name: string) => {
  const syntaxList = sourceFile.getChildrenOfKind(SyntaxKind.SyntaxList)[0];
  if (!syntaxList) {
    throw new Error('No SyntaxList');
  }
  const declaration = findDeclaration(syntaxList, name);
  if (!declaration) {
    throw new Error('No rawModels declaration');
  }
  return declaration;
};

const findDeclaration = (syntaxList: SyntaxList, name: string) => {
  for (const variableStatement of syntaxList.getChildrenOfKind(SyntaxKind.VariableStatement)) {
    for (const declaration of variableStatement.getDeclarationList().getDeclarations()) {
      if (declaration.getName() === name) {
        return declaration;
      }
    }
  }
};
