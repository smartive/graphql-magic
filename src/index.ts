// created from 'create-ts-index'

export * from './api';
export * from './client';
export * from './db';
export * from './migrations';
export * from './models';
export * from './permissions';
export * from './resolvers';
export * from './schema';
export * from './utils';
export * from './context';
export * from './errors';
export * from './values';
