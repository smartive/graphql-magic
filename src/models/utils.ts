import assert from 'assert';
import camelCase from 'lodash/camelCase';
import lodashGet from 'lodash/get';
import startCase from 'lodash/startCase';
import {
  CustomField,
  EntityField,
  EntityModel,
  EnumField,
  EnumModel,
  InputModel,
  InterfaceModel,
  Model,
  ObjectModel,
  PrimitiveField,
  RawEnumModel,
  RelationField,
  ScalarModel,
} from './models';

const isNotFalsy = <T>(v: T | null | undefined | false): v is T => typeof v !== 'undefined' && v !== null && v !== false;

export const merge = <T>(objects: (Record<string, T> | undefined | false)[] | undefined): Record<string, T> =>
  (objects || []).filter(isNotFalsy).reduce((i, acc) => ({ ...acc, ...i }), {});

// Target -> target
export const typeToField = (type: string) => type.substr(0, 1).toLowerCase() + type.substr(1);

export const getLabel = (s: string) => startCase(camelCase(s));

export const or =
  <T>(...predicates: ((field: T) => boolean)[]) =>
  (field: T) =>
    predicates.some((predicate) => predicate(field));

export const and =
  <T>(...predicates: ((field: T) => boolean)[]) =>
  (field: T) =>
    predicates.every((predicate) => predicate(field));

export const not =
  <T>(predicate: (field: T) => boolean) =>
  (field: T) =>
    !predicate(field);

export const isRootModel = (model: EntityModel) => !!model.root;

export const isEntityModel = (model: Model): model is EntityModel => model instanceof EntityModel;

export const isEnumModel = (model: Model): model is EnumModel => model instanceof EnumModel;

export const isRawEnumModel = (model: Model): model is RawEnumModel => model instanceof RawEnumModel;

export const isScalarModel = (model: Model): model is ScalarModel => model instanceof ScalarModel;

export const isObjectModel = (model: Model): model is ObjectModel => model instanceof ObjectModel;

export const isInputModel = (model: Model): model is InputModel => model instanceof InputModel;

export const isInterfaceModel = (model: Model): model is InterfaceModel => model instanceof InterfaceModel;

export const isCreatableModel = (model: EntityModel) => model.creatable && model.fields.some(isCreatableField);

export const isUpdatableModel = (model: EntityModel) => model.updatable && model.fields.some(isUpdatableField);

export const isCreatableField = (field: EntityField) => !field.inherited && !!field.creatable;

export const isUpdatableField = (field: EntityField) => !field.inherited && !!field.updatable;

export const modelNeedsTable = (model: EntityModel) => model.fields.some((field) => !field.inherited);

export const hasName = (name: string) => (field: EntityField) => field.name == name;

export const isPrimitive = (field: EntityField): field is PrimitiveField =>
  field.kind === undefined || field.kind === 'primitive';

export const isEnum = (field: EntityField): field is EnumField => field.kind === 'enum';

export const isRelation = (field: EntityField): field is RelationField => field.kind === 'relation';

export const isInherited = (field: EntityField) => !!field.inherited;

export const isInTable = (field: EntityField) => field.name === 'id' || !field.inherited;

export const isToOneRelation = (field: EntityField): field is RelationField => isRelation(field) && !!field.toOne;

export const isQueriableField = ({ queriable }: EntityField) => queriable !== false;

export const isCustomField = (field: EntityField): field is CustomField => field.kind === 'custom';

export const isVisible = ({ hidden }: EntityField) => hidden !== true;

export const isSimpleField = and(not(isRelation), not(isCustomField));

export const isUpdatable = ({ updatable }: EntityField) => !!updatable;

export const isCreatable = ({ creatable }: EntityField) => !!creatable;

export const isQueriableBy = (role: string) => (field: EntityField) =>
  field.queriable !== false &&
  (field.queriable === undefined ||
    field.queriable === true ||
    !field.queriable.roles ||
    field.queriable.roles.includes(role));

export const isUpdatableBy = (role: string) => (field: EntityField) =>
  field.updatable && (field.updatable === true || !field.updatable.roles || field.updatable.roles.includes(role));

export const isCreatableBy = (role: string) => (field: EntityField) =>
  field.creatable && (field.creatable === true || !field.creatable.roles || field.creatable.roles.includes(role));

export const getActionableRelations = (model: EntityModel, action: 'create' | 'update' | 'filter') =>
  model.relations
    .filter(
      (relation) =>
        relation.field[
          `${action === 'filter' ? action : action.slice(0, -1)}able` as 'filterable' | 'creatable' | 'updatable'
        ],
    )
    .map(({ name }) => name);

export const summonByName = <T extends { name: string }>(array: T[], value: string) => summonByKey(array, 'name', value);

export const summonByKey = <T>(array: readonly T[] | undefined, key: string, value: unknown) =>
  summon(array, (element: T) => lodashGet(element, key) === value, `No element found with ${key} ${value}`);

export const summon = <T>(array: readonly T[] | undefined, cb: Parameters<T[]['find']>[1], errorMessage?: string) => {
  if (array === undefined) {
    console.trace();
    throw new Error('Base array is not defined.');
  }
  const result = array.find(cb);
  if (result === undefined) {
    console.trace();
    throw new Error(errorMessage || 'Element not found.');
  }

  return result;
};

type ForSure<T> = T extends undefined | null ? never : T;

export const it = <T>(object: T | null | undefined): ForSure<T> => {
  if (object === undefined || object === null) {
    console.trace();
    throw new Error('Base object is not defined.');
  }

  return object as ForSure<T>;
};

export const get = <T, U extends keyof ForSure<T>>(object: T | null | undefined, key: U): ForSure<ForSure<T>[U]> => {
  const value = it(object)[key];
  if (value === undefined || value === null) {
    const error = new Error(`Object doesn't have ${String(key)}`);
    console.warn(error);
    throw error;
  }

  return value as ForSure<ForSure<T>[U]>;
};

export const getString = (v: unknown) => {
  assert(typeof v === 'string');

  return v;
};

export const retry = async <T>(cb: () => Promise<T>, condition: (e: any) => boolean) => {
  try {
    return await cb();
  } catch (e) {
    if (condition(e)) {
      return await cb();
    }
    throw e;
  }
};

type Typeof = {
  string: string;
  number: number;
  bigint: bigint;
  boolean: boolean;
  symbol: symbol;
  undefined: undefined;
  object: object;
  function: Function;
};

export const as = <T extends keyof Typeof>(value: unknown, type: T): Typeof[T] => {
  if (typeof value !== type) {
    throw new Error(`No string`);
  }

  return value as Typeof[T];
};
