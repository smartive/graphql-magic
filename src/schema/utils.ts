import { Dayjs } from 'dayjs';
import {
  ArgumentNode,
  DefinitionNode,
  DirectiveDefinitionNode,
  DirectiveNode,
  DocumentNode,
  EnumTypeDefinitionNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  InputObjectTypeDefinitionNode,
  InputValueDefinitionNode,
  InterfaceTypeDefinitionNode,
  ListTypeNode,
  NamedTypeNode,
  NameNode,
  NonNullTypeNode,
  ObjectFieldNode,
  ObjectTypeDefinitionNode,
  ScalarTypeDefinitionNode,
  StringValueNode,
  TypeNode,
  ValueNode,
} from 'graphql';
import { DateTime } from 'luxon';
import { Directive, Value, Values } from '../values';

export type Field = {
  name: string;
  type: string;
  description?: string;
  list?: boolean;
  nonNull?: boolean;
  defaultValue?: Value;
  args?: Field[];
  directives?: Directive[];
};

export type DirectiveLocation =
  | 'ARGUMENT_DEFINITION'
  | 'INPUT_FIELD_DEFINITION'
  | 'FIELD'
  | 'FIELD_DEFINITION'
  | 'OBJECT'
  | 'INTERFACE'
  | 'INPUT_OBJECT';

export const document = (definitions: DefinitionNode[]): DocumentNode => ({
  kind: 'Document',
  definitions,
});

export const directive = (nme: string, locations: DirectiveLocation[], fields?: Field[]): DirectiveDefinitionNode => ({
  name: name(nme),
  repeatable: false,
  kind: 'DirectiveDefinition',
  arguments: fields && inputValues(fields),
  locations: locations.map(name),
});

export const scalar = (nme: string): ScalarTypeDefinitionNode => ({
  name: name(nme),
  kind: 'ScalarTypeDefinition',
});

export const input = (nme: string, fields: Field[], dvs?: Directive[]): InputObjectTypeDefinitionNode => ({
  name: name(nme),
  fields: inputValues(fields),
  kind: 'InputObjectTypeDefinition',
  directives: directives(dvs),
});

export const object = (nme: string, fds: Field[], interfaces?: string[], dvs?: Directive[]): ObjectTypeDefinitionNode => ({
  name: name(nme),
  fields: fields(fds),
  kind: 'ObjectTypeDefinition',
  interfaces: interfaces?.map((i) => namedType(i)),
  directives: directives(dvs),
});

export const iface = (nme: string, fds: Field[], interfaces?: string[], dvs?: Directive[]): InterfaceTypeDefinitionNode => ({
  name: name(nme),
  fields: fields(fds),
  kind: 'InterfaceTypeDefinition',
  interfaces: interfaces?.map((i) => namedType(i)),
  directives: directives(dvs),
});

export const inputValues = (fields: Field[]): InputValueDefinitionNode[] =>
  fields.map(
    (field): InputValueDefinitionNode => ({
      kind: 'InputValueDefinition',
      name: name(field.name),
      type: fieldType(field),
      defaultValue: field.defaultValue === undefined ? undefined : value(field.defaultValue),
      directives: directives(field.directives),
    }),
  );

export const fields = (fields: Field[]): FieldDefinitionNode[] =>
  fields.map(
    (field): FieldDefinitionNode => ({
      kind: 'FieldDefinition',
      name: name(field.name),
      type: fieldType(field),
      arguments: field.args?.map((arg) => ({
        kind: 'InputValueDefinition',
        name: name(arg.name),
        type: fieldType(arg),
        defaultValue: arg.defaultValue === undefined ? undefined : value(arg.defaultValue),
      })),
      directives: directives(field.directives),
    }),
  );

export const inputValue = (
  nme: string,
  type: TypeNode,
  dvs?: Directive[],
  description?: string,
  defaultValue?: Value,
): InputValueDefinitionNode => ({
  name: name(nme),
  type,
  kind: 'InputValueDefinition',
  directives: directives(dvs),
  defaultValue: defaultValue ? value(defaultValue) : undefined,
  description: description ? (value(description) as StringValueNode) : undefined,
});

export const directives = (directives?: Directive[]): DirectiveNode[] | undefined =>
  directives?.map(
    (directive): DirectiveNode => ({
      kind: 'Directive',
      name: name(directive.name),
      arguments: args(directive.values),
    }),
  );

export const args = (ags: Values | undefined): readonly ArgumentNode[] | undefined =>
  ags?.map(
    (argument): ArgumentNode => ({
      kind: 'Argument',
      name: name(argument.name),
      value: value(argument.values),
    }),
  );

export const enm = (nme: string, values: string[]): EnumTypeDefinitionNode => ({
  name: name(nme),
  kind: 'EnumTypeDefinition',
  values: values.map(
    (v): EnumValueDefinitionNode => ({
      kind: 'EnumValueDefinition',
      name: name(v),
    }),
  ),
});

export const nonNull = (type: NamedTypeNode | ListTypeNode): NonNullTypeNode => ({
  type,
  kind: 'NonNullType',
});

export const namedType = (nme: string): NamedTypeNode => ({
  kind: 'NamedType',
  name: name(nme),
});

export const list = (type: TypeNode): ListTypeNode => ({
  type,
  kind: 'ListType',
});

export const name = (name: string): NameNode => ({
  kind: 'Name',
  value: name,
});

export const fieldType = (field: Field) => {
  let type: TypeNode = namedType(field.type);
  if (field.list) {
    type = list(nonNull(type));
  }
  if (field.nonNull) {
    type = nonNull(type);
  }

  return type;
};

export const value = (val: Value = null): ValueNode =>
  val === null
    ? {
        kind: 'NullValue',
      }
    : typeof val === 'boolean'
      ? {
          kind: 'BooleanValue',
          value: val,
        }
      : typeof val === 'number'
        ? {
            kind: 'IntValue',
            value: `${val}`,
          }
        : typeof val === 'string'
          ? {
              kind: 'StringValue',
              value: val,
            }
          : val instanceof Symbol
            ? {
                kind: 'EnumValue',
                value: val.description!,
              }
            : Array.isArray(val)
              ? {
                  kind: 'ListValue',
                  values: val.map(value),
                }
              : val instanceof DateTime
                ? {
                    kind: 'StringValue',
                    value: val.toString(),
                  }
                : val instanceof Dayjs
                  ? {
                      kind: 'StringValue',
                      value: val.toISOString(),
                    }
                  : typeof val === 'object'
                    ? {
                        kind: 'ObjectValue',
                        fields: Object.keys(val).map(
                          (nme): ObjectFieldNode => ({
                            kind: 'ObjectField',
                            name: name(nme),
                            value: value(val[nme]),
                          }),
                        ),
                      }
                    : doThrow(`Unsupported value ${val}`);

const doThrow = (message: string) => {
  throw new Error(message);
};
