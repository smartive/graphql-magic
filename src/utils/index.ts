// created from 'create-ts-index'

export * from './dates';
export * from './rules';
