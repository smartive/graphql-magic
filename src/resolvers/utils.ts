import { createHash } from 'crypto';
import type {
  ASTNode,
  FieldNode,
  FragmentDefinitionNode,
  FragmentSpreadNode,
  GraphQLObjectType,
  GraphQLSchema,
  InlineFragmentNode,
  TypeNode,
} from 'graphql';
import { Kind } from 'graphql';
import { Knex } from 'knex';
import { isEqual } from 'lodash';
import { EntityField } from '..';
import { UserInputError } from '../errors';
import { get, it, summon } from '../models/utils';
import { Value } from '../values';
import { FieldResolverNode, ResolverNode } from './node';

export const ID_ALIAS = 'ID';
export const TYPE_ALIAS = 'TYPE';

export type VariableValues = Record<string, Value>;

export const getTypeName = (t: TypeNode): string => {
  switch (t.kind) {
    case 'ListType':
    case 'NonNullType':
      return getTypeName(t.type);
    case 'NamedType':
      return t.name.value;
    default:
      throw new Error(`Unknown type node kind: ${t.kind}`);
  }
};

export const isListType = (type: TypeNode): boolean => {
  switch (type.kind) {
    case 'ListType':
      return true;
    case 'NonNullType':
      return isListType(type.type);
    default:
      return false;
  }
};

export const isFieldNode = (n: ASTNode): n is FieldNode => n.kind === Kind.FIELD;

export const isInlineFragmentNode = (n: ASTNode): n is InlineFragmentNode => n.kind === Kind.INLINE_FRAGMENT;

export const isFragmentSpreadNode = (n: ASTNode): n is FragmentSpreadNode => n.kind === Kind.FRAGMENT_SPREAD;

export type Maybe<T> = null | undefined | T;

export const getType = (schema: GraphQLSchema, name: string) =>
  get(schema.getType(name) as Maybe<GraphQLObjectType>, 'astNode');

export const getFragmentTypeName = (node: InlineFragmentNode | FragmentDefinitionNode) =>
  get(get(node.typeCondition, 'name'), 'value');

export type Entry = {
  [ID_ALIAS]: string;
  [field: string]: null | string | number | Entry | Entry[];
};

export function hydrate<T extends Entry>(
  node: FieldResolverNode,
  raw: Record<string, undefined | null | string | Date | number>[],
): T[] {
  const resultAlias = node.resultAlias;
  const res = raw.map((entry) => {
    const res: any = {};
    outer: for (const [column, value] of Object.entries(entry)) {
      let current = res;
      const [, columnWithoutField, fieldName] = column.match(/^(.*\w)__(\w+)$/)!;
      const longColumn = node.ctx.aliases.getLong(columnWithoutField);
      const longColumnWithoutRoot = longColumn.replace(new RegExp(`^${resultAlias}(__)?`), '');
      const allParts = [resultAlias, ...(longColumnWithoutRoot ? longColumnWithoutRoot.split('__') : []), fieldName];
      for (let i = 0; i < allParts.length - 1; i++) {
        const part = allParts[i];

        if (!current[part]) {
          const idField = [node.ctx.aliases.getShort(allParts.slice(0, i + 1).join('__')), ID_ALIAS].join('__');
          if (!entry[idField]) {
            continue outer;
          }
          current[part] = {};
        }
        current = current[part];
      }
      current[it(fieldName)] = value;
    }

    return res[resultAlias];
  });

  return res;
}

export const ors = (query: Knex.QueryBuilder, [first, ...rest]: ((query: Knex.QueryBuilder) => Knex.QueryBuilder)[]) => {
  if (!first) {
    return query;
  }

  return query.where((subQuery) => {
    void subQuery.where((subSubQuery) => {
      void first(subSubQuery);
    });
    for (const cb of rest) {
      void subQuery.orWhere((subSubQuery) => {
        void cb(subSubQuery);
      });
    }
  });
};

export const getNameOrAlias = (node: FieldNode) => {
  const name = node.alias ? node.alias.value : node.name.value;

  if ([ID_ALIAS].includes(name)) {
    throw new UserInputError(`"${name}" can not be used as alias since it's a reserved word`);
  }

  return name;
};

export type Ops<T> = ((target: T) => T)[];

export type QueryBuilderOps = Ops<Knex.QueryBuilder>;

export const apply = <T>(target: T, ops: ((target: T) => T)[]) => ops.reduce((target, op) => op(target), target);

type Join = { table1Alias: string; column1: string; table2Name: string; table2Alias: string; column2: string };
export type Joins = Join[];

export const applyJoins = (aliases: AliasGenerator, query: Knex.QueryBuilder, joins: Joins) => {
  for (const { table1Alias, table2Name, table2Alias, column1, column2 } of joins) {
    const table1ShortAlias = aliases.getShort(table1Alias);
    const table2ShortAlias = aliases.getShort(table2Alias);
    void query.leftJoin(
      `${table2Name} as ${table2ShortAlias}`,
      `${table1ShortAlias}.${column1}`,
      `${table2ShortAlias}.${column2}`,
    );
  }
};

/**
 * Note: modifies the join parameter.
 */
export const addJoin = (
  joins: Joins,
  table1Alias: string,
  table2Name: string,
  table2Alias: string,
  column1: string,
  column2: string,
) => {
  const join = { table1Alias, table2Name, table2Alias, column1, column2 };
  const existingJoin = joins.find((j) => j.table2Alias === join.table2Alias);
  if (existingJoin) {
    if (!isEqual(existingJoin, join)) {
      throw new Error(`Join collision: ${existingJoin}, ${join}`);
    }
  } else {
    joins.push(join);
  }
};

export class AliasGenerator {
  private reverse: Record<string, string> = {};

  public getShort(long?: string) {
    if (!long) {
      const short = `a${Object.keys(this.reverse).length}`;

      return (this.reverse[short] = short);
    }

    const shortPrefix = long
      .split('__')
      .map((part) => part[0] + part.slice(1).replaceAll(/[a-z]+/g, ''))
      .join('__');
    let postfix = 0;
    let short = shortPrefix + (postfix || '');
    while (short in this.reverse && this.reverse[short] !== long) {
      short = shortPrefix + ++postfix;
    }

    this.reverse[short] = long;

    return short;
  }

  public getLong(short: string) {
    return get(this.reverse, short);
  }
}

export const hash = (s: any) => createHash('md5').update(JSON.stringify(s)).digest('hex');

export const getColumnName = (field: EntityField) =>
  field.kind === 'relation' ? field.foreignKey || `${field.name}Id` : field.name;

export const getColumn = (node: Pick<ResolverNode, 'model' | 'ctx' | 'rootTableAlias' | 'tableAlias'>, key: string) => {
  const field = summon(node.model.fields, (field) => getColumnName(field) === key);

  return `${node.ctx.aliases.getShort(field.inherited ? node.rootTableAlias : node.tableAlias)}.${key}`;
};
