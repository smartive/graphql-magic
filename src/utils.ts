import assert from 'assert';
import { pluralize } from 'inflection';
import { camelCase, kebabCase, get as lodashGet, startCase } from 'lodash';
import { Model, Models, ObjectModel, RawModels, Relation, ReverseRelation, isObjectModel } from './models';

const isNotFalsy = <T>(v: T | null | undefined | false): v is T => typeof v !== 'undefined' && v !== null && v !== false;

export const merge = <T>(objects: ({ [name: string]: T } | undefined | false)[] | undefined): { [name: string]: T } =>
  (objects || []).filter(isNotFalsy).reduce((i, acc) => ({ ...acc, ...i }), {});

// Target -> target
export const typeToField = (type: string) => type.substr(0, 1).toLowerCase() + type.substr(1);

export const getModelPlural = (model: ObjectModel | Model) => model.plural || pluralize(model.name);

export const getModelPluralField = (model: Model) => typeToField(getModelPlural(model));

export const getModelSlug = (model: Model) => kebabCase(getModelPlural(model));

export const getModelLabelPlural = (model: Model) => getLabel(getModelPlural(model));

export const getModelLabel = (model: Model) => getLabel(model.name);

export const getLabel = (s: string) => startCase(camelCase(s));

export const getModels = (rawModels: RawModels): Models => {
  const models: Models = rawModels.filter(isObjectModel).map((model) => {
    const objectModel: Model = {
      ...model,
      fieldsByName: {},
      relations: [],
      relationsByName: {},
      reverseRelations: [],
      reverseRelationsByName: {},
      fields: [
        { name: 'id', type: 'ID', nonNull: true, unique: true, primary: true, generated: true },
        ...model.fields,
        ...(model.creatable
          ? [
              { name: 'createdAt', type: 'DateTime', nonNull: !model.nonStrict, orderable: true, generated: true },
              {
                name: 'createdBy',
                type: 'User',
                relation: true,
                nonNull: !model.nonStrict,
                reverse: `created${getModelPlural(model)}`,
                generated: true,
              },
            ]
          : []),
        ...(model.updatable
          ? [
              { name: 'updatedAt', type: 'DateTime', nonNull: !model.nonStrict, orderable: true, generated: true },
              {
                name: 'updatedBy',
                type: 'User',
                relation: true,
                nonNull: !model.nonStrict,
                reverse: `updated${getModelPlural(model)}`,
                generated: true,
              },
            ]
          : []),
        ...(model.deletable
          ? [
              {
                name: 'deleted',
                type: 'Boolean',
                nonNull: true,
                default: false,
                filterable: true,
                defaultFilter: false,
                generated: true,
              },
              { name: 'deletedAt', type: 'DateTime', orderable: true, generated: true },
              {
                name: 'deletedBy',
                type: 'User',
                relation: true,
                reverse: `deleted${getModelPlural(model)}`,
                generated: true,
              },
            ]
          : []),
      ].map(({ foreignKey, ...field }) => ({
        ...field,
        ...(field.relation && {
          foreignKey: foreignKey || `${field.name}Id`,
        }),
      })),
    };

    for (const field of objectModel.fields) {
      objectModel.fieldsByName[field.name] = field;
    }

    return objectModel;
  });

  for (const model of models) {
    for (const field of model.fields.filter(({ relation }) => relation)) {
      const fieldModel = summonByName(models, field.type);

      const reverseRelation: ReverseRelation = {
        name: field.reverse || (field.toOne ? typeToField(model.name) : getModelPluralField(model)),
        foreignKey: get(field, 'foreignKey'),
        type: model.name,
        toOne: !!field.toOne,
        fieldModel,
        field,
        model,
      };

      const relation: Relation = {
        field,
        model: fieldModel,
        reverseRelation,
      };
      model.relations.push(relation);
      model.relationsByName[relation.field.name] = relation;

      fieldModel.reverseRelations.push(reverseRelation);

      fieldModel.reverseRelationsByName[reverseRelation.name] = reverseRelation;
    }
  }

  return models;
};

export const summonByName = <T extends { name: string }>(array: T[], value: string) => summonByKey(array, 'name', value);

export const summonByKey = <T, U extends keyof T>(array: readonly T[] | undefined, key: string, value: unknown) =>
  summon(array, (element: T) => lodashGet(element, key) === value, `No element found with ${key} ${value}`);

export const summon = <T>(array: readonly T[] | undefined, cb: Parameters<T[]['find']>[1], errorMessage?: string) => {
  if (array === undefined) {
    throw new Error('Base array is not defined.');
  }
  const result = array.find(cb);
  if (result === undefined) {
    console.trace();
    throw new Error(errorMessage || 'Element not found.');
  }
  return result;
};

type ForSure<T> = T extends undefined | null ? never : T;

export const it = <T>(object: T | null | undefined): ForSure<T> => {
  if (object === undefined || object === null) {
    console.trace();
    throw new Error('Base object is not defined.');
  }

  return object as ForSure<T>;
};

export const get = <T, U extends keyof ForSure<T>>(object: T | null | undefined, key: U): ForSure<ForSure<T>[U]> => {
  const value = it(object)[key];
  if (value === undefined || value === null) {
    console.trace();
    throw new Error(`Object doesn't have ${String(key)}`);
  }
  return value as ForSure<ForSure<T>[U]>;
};

export const getString = (v: unknown) => {
  assert(typeof v === 'string');
  return v;
};

export const retry = async <T>(cb: () => Promise<T>, condition: (e: any) => boolean) => {
  try {
    return await cb();
  } catch (e) {
    if (condition(e)) {
      return await cb();
    } else {
      throw e;
    }
  }
};
