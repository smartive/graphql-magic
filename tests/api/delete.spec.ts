import { gql } from '../../src';
import { ANOTHER_ID } from '../utils/database/seed';
import { withServer } from '../utils/server';

describe('delete', () => {
  it('works with self-referential entities', async () => {
    await withServer(async (request) => {
      expect(
        await request(gql`
          mutation DeleteAnotherObject {
            deleteAnotherObject(where: { id: "${ANOTHER_ID}" })
          }
        `)
      ).toMatchSnapshot();

      expect(
        await request(gql`
          query GetAnotherObject {
            anotherObjects(where: { id: "${ANOTHER_ID}", deleted: true }) {
              id
              deleted
            }
          }
        `)
      ).toMatchSnapshot();
    });
  });
});
