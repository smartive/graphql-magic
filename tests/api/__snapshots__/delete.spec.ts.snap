// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete works with self-referential entities 1`] = `
{
  "deleteAnotherObject": "226a20e8-5c18-4423-99ca-eb0df6ff4fdd",
}
`;

exports[`delete works with self-referential entities 2`] = `
{
  "anotherObjects": [
    {
      "deleted": true,
      "id": "226a20e8-5c18-4423-99ca-eb0df6ff4fdd",
    },
  ],
}
`;
